admin/aaxplaymidi.1.in
.\" Manpage for aaxplay.
.\" Contact tech@adalin.com to correct errors or typos.
.TH man 1 "25 Sep 2018" "@AAX_UTILS_MAJOR_VERSION@.@AAX_UTILS_MINOR_VERSION@.@AAX_UTILS_MICRO_VERSION@" "aaxplaymidi man page"
.SH NAME
aaxplay \- Plays a midi file to an audio output device.
.SH SYNOPSIS
.B aaxplaymidi
[\fIOPTION\fR]...
.SH DESCRIPTION
.PP
Plays a midi file to an audio output device.
.TP
\fB\-i\fR, \fB\-\-input \fRFILE\fR
playback audio from a file
.TP
\fB\-d\fR, \fB\-\-device \fRDEVICE\fR
playback device (default if not specified)
.TP
\fB\-v\fR, \fB\-\-verbose
show extra playback information
.TP
\fB\-h\fR, \fB\-\-help
print this message and exit
.PP
Written by Erik Hofman <tech@adalin.com>
.SH SEE ALSO
aaxplay(1), aaxinfo(1)



ChangeLog
* Sat Oct 13 2018 - tech@adalin.org
- AeonWave (3.5.1) stable; urgency=low

 * Improve various instrument configurations.
 * Fix some more issues with the timed gain filter and timed pitch effect.
 * Remove a debugging statement.

-- Erik Hofman <tech@adalin.org>  Sat 13 oct  2018 10:37:00 + 0200

* Thu Sep 27 2018 - tech@adalin.org
- AeonWave (3.5.0) stable; urgency=low

 * Add the option to set the location of the shared-data directory at runtime
   using the new aaxDriverSetSetup function.
 * Add a C++ instrument header file which can be used to simulate a single,
   complete, musical instrument. And use it for the MIDI playback code.
 * Re-initialize the emitter after it is set to suspended. This makes the
   emitter ready for the next playing state.
 * Fix a possible race condition when using tied properties.
 * Fix some issues with the timed gain filter and timed pitch effect.
 * Fix an issue when two registered audio-frames both use a delay effect.
 * Add the option to define the loop staring point and loop end point in AAXS
   files that include a file in the sound section.
 * Allow AAXS defined filters and effects to be set as optional, in which case
   they will be omitted on systems which are low on resources (CPU and memory).
 * AAXS generated buffers will now use mipmapping to extend the pitch range
   beyond a pitch level of 32.

-- Erik Hofman <tech@adalin.org>  Thu 27 Sep  2018 09:13:00 + 0200

* Wed Aug 01 2018 - tech@adalin.org
- AeonWave (3.4.0) stable; urgency=low

 * Add support for the ISO9613-1 specified distance attenuation model.
 * If the ISO9613-1 distance  model is applied, and the velocity effect was not
   yet set, then the corresponding speed of sound will be set automatically.
 * If the source has stopped then finish the current time envelope and stop,
   otherwise an evelope with repeat set to max will never stop until the
   emitter is set to processed.
 * Add a RingModulator effect, mainly to be able to simulate atmospheric
   anomalies for wireless communications. But useful for other purposes too.
 * Remove debugging output from ogg streaming code.

-- Erik Hofman <tech@adalin.org>  Wed 01 Aug  2018 11:00:00 + 0200

* Fri Jul 20 2018 - tech@adalin.org
- AeonWave (3.3.0) stable; urgency=low

 * Automatically convert an effect name like "effects/nature/birds" to a full
   path name using APP_DATA_DIR when it is parsed to the buffer retrieving
   function of the AeonWave class.
 * Add a Param class which encapsulates a float variable and which can
   be tied to a DSP parameter. Every time the param gets a new value
   assigned the DSP parameter will be updated too.
 * Fix a small problem with stopping an emitter which has a timed-gain envelope
   defined with one of the time settings to "inf". They should now finish
   properly after the emitter is requested to stop playing.
 * Overhaul of the C++ classes to make them C++11 compliant.
 * Remove support for an ancient way of handling config files.
 * Prevent a crash when the state of the timed-gain filter or timed-pitch effect
   is not set before it is registered the sensor, audioframe or emitter.

-- Erik Hofman <tech@adalin.org>  Fri 20 Jul  2018 14:01:00 + 0200

* Tue Jul 03 2018 - tech@adalin.org
- AeonWave (3.2.3) stable; urgency=low
 * Fix a missing offset(enum aaxType) error for the AeonWave class.
 * Fix a problem with AAX_SHARED_DATA_DIR support for Windows.

-- Erik Hofman <tech@adalin.org>  Tue 03 Jul  2018 08:49:00 + 0200

* Sat Jun 30 2018 - tech@adalin.org
- AeonWave (3.2.2) stable; urgency=medium
 * Add the AAX_PCM24S_PACKED format for 3-byte packed 24-bit samples.
   AAX_PCM24S is used for 32-bit encoded 24-bit samples.
 * Fix 24-bit packed PCM data handling.
 * Fix a possible stream synchronization problem at the start of streams
   that support ICY metadata song information.

-- Erik Hofman <tech@adalin.org>  Sat 30 Jun  2018 11:06:00 + 0200

* Sat Jun 23 2018 - tech@adalin.org
- AeonWave (3.2.1) stable; urgency=low
 * Improve timing for some backends by using a regular interval method
   instead of waiting just a number of milliseconds from the time of calling.
   This moslty affected OSS and SDL backends.
 * Move the mixing code from the callback to the backend-thread for the SDL
   backend reducing the time spent in the callers code to a mere memcopy.
 * Enable the SDL directsound backend for Windows XP.
 * Rename the angular filter (which was lost in translation) to directional
   filter. For backwards compatibility the angular name is still being
   recognized internally and the library is binary compatible with previous
   versions.

-- Erik Hofman <tech@adalin.org>  Sat 23 Jun  2018 12:31:00 + 0200

* Sat Jun 16 2018 - tech@adalin.org
- AeonWave (3.2.0) stable; urgency=high
 * Add SDL audio support for backends AeonWave does not support natively.
   This potentially opens up the option to use AeonWave on MacOS and BSD.
   SDL 2.0.5 or later is required.
 * Allow retrieving the data directory which AeonWave uses to store shared data
   like preset files by calling aaxDriverGetSetup(cfg, AAX_SHARED_DATA_DIR)
 * Allow defining AAX_ENVELOPE_FOLLOW or AAX_INVERSE_ENVELOPE_FOLLOW alongside
   the waveform type to control the noise behavior when setting the state for
   the bitcrusher filter. The default behavior is still a constant noise level.
 * Fix a number of potential buffer overflows when enumerating the available
   backend interfaces.
 * Fix a number of one-off memory leaks.

-- Erik Hofman <tech@adalin.org>  Sat 16 Jun  2018 13:30:00 + 0200

* Tue Jun 05 2018 - tech@adalin.org
- AeonWave (3.1.2) stable; urgency=low
 * Add the bitcrusher filter to simulate (communication line) sound degradation.
 * Delever on the promis that one can add a named, but empty, buffer in
   case aaxBufferReadFromStream fails when calling buffer() from the
   aeonwave class.

-- Erik Hofman <tech@adalin.org>  Tue 05 Jun  2018 20:48:00 + 0200

* Fri Jun 01 2018 - tech@adalin.org

src/aaxplaymidi.cpp
/*
 *  MIT License
 * 
 * Copyright (c) 2018 
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// Inspired by: https://github.com/aguaviva/SimpleMidi

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <assert.h>
#ifdef HAVE_RMALLOC_H
# include <rmalloc.h>
#else
# include <string.h>
# if HAVE_STRINGS_H
#  include <strings.h>
# endif
#endif

#include <aax/aeonwave.hpp>
#include <aax/instrument.hpp>

#include <midi/midi.hpp>

#include "driver.h"

#define IFILE_PATH		SRC_PATH"/beethoven_opus10_3.mid"
#define INSTRUMENT		"instruments/piano-accoustic"
#define INSTRUMENTS		"gmmidi.xml"
#define DRUMS			"gmdrums.xml"

void
help()
{
    aaxConfig cfgi, cfgo;

    printf("aaxplaymidi version %i.%i.%i\n\n", AAX_UTILS_MAJOR_VERSION,
                                           AAX_UTILS_MINOR_VERSION,
                                           AAX_UTILS_MICRO_VERSION);
    printf("Usage: aaxplaymidi [options]\n");
    printf("Plays a MIDI file to an audio output device.\n");

    printf("\nOptions:\n");
    printf("  -i, --input <file>\t\tplayback audio from a file\n");
    printf("  -d, --device <device>\t\tplayback device (default if not specified)\n");
    printf("  -v, --verbose\t\t\tshow extra playback information\n");
    printf("  -h, --help\t\t\tprint this message and exit\n");

    printf("\nUse aaxplay for playing other audio file formats.\n");

    printf("\n");

    exit(-1);
}

int verbose = 0;
int main(int argc, char **argv)
{
    if (argc == 1 || getCommandLineOption(argc, argv, "-h") ||
                     getCommandLineOption(argc, argv, "--help"))
    {
        help();
    }

    char *devname = getDeviceName(argc, argv);
    char *infile = getInputFile(argc, argv, IFILE_PATH);
    try
    {
        aax::MIDIFile midi(devname, infile);
        if (midi)
        {
            int64_t sleep_us, dt_us;
            uint64_t time_parts = 0;
            uint32_t wait_parts;
            struct timeval now;

            if (getCommandLineOption(argc, argv, "-v") ||
                getCommandLineOption(argc, argv, "--verbose"))
            {
               midi.set_verbose(true);
            }

#ifndef NDEBUG
            midi.set_file_path(SRC_PATH"/midi");
            midi.set_instrument_file(INSTRUMENTS);
            midi.set_drum_file(DRUMS);
#endif

            midi.initialize();
            midi.start();

            wait_parts = 1000;
            set_mode(1);

            gettimeofday(&now, NULL);
            dt_us = -(now.tv_sec * 1000000 + now.tv_usec);
            do
            {
                if (!midi.process(time_parts, wait_parts)) break;

                if (wait_parts > 0)
                {
                    uint32_t wait_us;

                    gettimeofday(&now, NULL);
                    dt_us += now.tv_sec * 1000000 + now.tv_usec;

                    wait_us = wait_parts*midi.get_uspp();
                    sleep_us = wait_us - dt_us;
                    if (sleep_us > 0) {
                       usecSleep(sleep_us);
                    }

                    gettimeofday(&now, NULL);
                    dt_us = -(now.tv_sec * 1000000 + now.tv_usec);

                    time_parts += wait_parts;
                }
            }
            while(!get_key());
            set_mode(0);

            midi.stop();
        }
    } catch (const std::exception& e) {
        std::cerr << "Error while processing the MIDI file: "
                  << e.what() << std::endl;
    }


    return 0;
}


